crates/items/src/item_stack.rs

```diff
@@ -0,0 +1,239 @@
+        self.count
+    }
+
+    /// Removes some items from this `ItemStack`.
+    pub fn remove(&mut self, count: u32) -> Result<u32, ()> {
+        if self.count < count {
+            return Err(());
+        }
+        self.count -= count;
+        Ok(self.count)
+    }
+
+    /// Sets the item type for this `ItemStack`. Returns the new
+    /// item type or fails if the current item count exceeds the
+    /// new item type stack size.
+    pub fn set_item(&mut self, item: Item) -> Result<Item, ()> {
+        if self.count > item.stack_size() {
+            return Err(());
+        }
+        self.item = item;
+        Ok(self.item)
+    }
+
+    /// Sets the item type for this `ItemStack`. Does not check if
+    /// the new item type stack size will be lower than the current
+    /// item count. Returns the new item type.
+    pub fn unchecked_set_item(&mut self, item: Item) -> Item {
+        self.item = item;
+        self.item
+    }
+
+    /// Sets the count for this `ItemStack`. Returns the updated
+    /// count or fails if the new count would exceed the stack
+    /// size for that item type.
+    pub fn set_count(&mut self, count: u32) -> Result<u32, ()> {
+        if count > self.item.stack_size() {
+            return Err(());
+        }
+        self.count = count;
+        Ok(self.count)
+    }
+
+    /// Sets the count for this `ItemStack`. It will not check if
+    /// the desired count exceeds the current item type stack size.
+    /// Returns the updated count.
+    pub fn unchecked_set_count(&mut self, count: u32) -> u32 {
+        self.count = count;
+        self.count
+    }
+
+    /// Splits this `ItemStack` in half, returning the
+    /// removed half. If the amount is odd, `self`
+    /// will be left with the least items. Returns the taken
+    /// half.
+    pub fn take_half(&mut self) -> ItemStack {
+        self.take((self.count as f64 / 2 as f64).ceil() as u32)
+            .unwrap()
+    }
+
+    /// Splits this `ItemStack` by removing the
+    /// specified amount. Returns the taken part.
+    pub fn take(&mut self, amount: u32) -> Result<ItemStack, ()> {
+        let count_left: i32 = self.count as i32 - amount as i32;
+        if count_left < 0 {
+            return Err(());
+        }
+        let taken = ItemStack {
+            count: amount,
+            ..self.clone()
+        };
+        self.count = count_left as u32;
+        Ok(taken)
+    }
+
+    /// Merges another `ItemStack` with this one.
+    pub fn merge_with(&mut self, other: &mut Self) -> bool {
+        if !self.has_same_type_and_damage(other) {
+            return false;
+        }
+        let new_count = (self.count + other.count).min(self.item.stack_size());
+        let amount_added = new_count - self.count;
+        self.count = new_count;
+        other.count -= amount_added;
+        true
+    }
+
+    /// Transfers up to `n` items to `other`.
+    pub fn transfer_to(&mut self, n: u32, other: &mut Self) {
```

Comment:
This also feels a bit weird, `self` can end up having `self.count = 0`

Reply:
Handled in next commit.